core/cp4d_authenticator.go

```diff
@@ -144,24 +151,80 @@ func (authenticator *CloudPakForDataAuthenticator) Authenticate(request *http.Re
 }
 
 // getToken: returns an access token to be used in an Authorization header.
-// Whenever a new token is needed (when a token doesn't yet exist, or the existing token has expired),
-// a new access token is fetched from the token server.
+// Whenever a new token is needed (when a token doesn't yet exist, needs to be refreshed,
+// or the existing token has expired), a new access token is fetched from the token server.
 func (authenticator *CloudPakForDataAuthenticator) getToken() (string, error) {
 	if authenticator.tokenData == nil || !authenticator.tokenData.isTokenValid() {
-		tokenResponse, err := authenticator.requestToken()
+		// synchronously request the token
+		err := authenticator.synchronizedRequestToken()
 		if err != nil {
 			return "", err
 		}
-
-		authenticator.tokenData, err = newCp4dTokenData(tokenResponse)
-		if err != nil {
-			return "", err
+	} else if authenticator.tokenData != nil && authenticator.needsRefresh() {
+		// If refresh needed, kick off a go routine in the background to get a new token
+		ch := make(chan error)
```

Comment:
I created a goroutine to achieve the "multi-threading" we want. The select case below it should ensure that no other routine sits around and waits for the newly spawned routine to finish